### DISCUSSION TITLE
pg_hba should also be copied to replicas when postgresql_exists=true

### DISCUSSION THREAD

**Discussion by tomo-kivra** [autobase-tech/autobase#92] [General]
Hi,
Very nice playbook, but I think when having an existing PG cluster we should still copy pg_hba.conf to replicas also.
Currently it is not done, and after failover is done to one of replicas, pg_hba will different than on previous master.
(of course one should prepare pg_hba to be the same as before, but this is in case we want to change something when deploying patroni)

A quick fix would be to add additional "OR" to when statement:

```
--- a/roles/patroni/tasks/main.yml
+++ b/roles/patroni/tasks/main.yml
@@ -901,8 +901,9 @@
         owner: postgres
         group: postgres
         mode: 0640
-  when: existing_pgcluster is defined and existing_pgcluster|bool
-        and postgresql_conf_dir != postgresql_data_dir
+  when: (existing_pgcluster is defined and existing_pgcluster|bool
+        and postgresql_conf_dir != postgresql_data_dir)
+        or postgresql_exists == "true"
   tags: patroni, pg_hba, pg_hba_generate
 
 - block:  # start patroni on replica

```

Maybe I am missing something here, so would like to see your comment on this.

**Comment by vitabaks** [comment]
pg_hba.conf  must be copied from master, with the add_pgnode.yml playbook.
What is your Linux distro?

**Reply by tomo-kivra** [reply]
It is CentOS 7.
If I understood correctly add_pgnode.yml is used to expand with additional node. In this case, I have existing postgresql cluster of 2 nodes.
Pg_hba is re-generated on primary, but it is not changed on replicas.

Again, just to point out, this happens only if I want to change pg_hba.conf while deploying patroni over existing cluster.

**Reply by vitabaks** [reply]
For Centos, there is no need to generate pg_hba.conf on replicas, since the configuration files are located in the PGDATA directory, which will be copied from the master to the replica server, using pg_basebackup.

For this corresponds to the condition `postgresql_conf_dir != postgresql_data_dir`

**Reply by vitabaks** [reply]
> Pg_hba is re-generated on primary, but it is not changed on replicas.

Do you want to change the pg_hba on all nodes of the current cluster using ansible playbook?

**Reply by tomo-kivra** [reply]
> For Centos, there is no need to generate pg_hba.conf on replicas, since the configuration files are located in the PGDATA directory, which will be copied from the master to the replica server, using pg_basebackup.

Yes, but only if I add new nodes. This will not happen for existing postgresql nodes.

> Do you want to change the pg_hba on all nodes of the current cluster using ansible playbook?

Yes, since now I will have pgbouncer in front of postgresql, so I want to have pgbouncer do authentication.

**Reply by vitabaks** [reply]
See commit 1887ffa22ceffbb9314098541b40e858d2c77319

**Reply by tomo-kivra** [reply]
Great. Thanks.
